Build an example CUDA app if CUDA driver is already built.

wangkuiyi opened the pull request:
Fixes https://github.com/k8sp/issues/issues/3

wangkuiyi committed:
Separate gcc49 installation out from build.sh

wangkuiyi committed:
Rename build.sh into buld-cuda-driver.sh

wangkuiyi committed:
Add functionality to build CUDA apps

wangkuiyi committed:
Update README.md

wangkuiyi committed:
Remove uncessary install-gcc49.sh, as we use Docker image gcc:4.9 now

lipeng-unisound commented:
When I ran the script, I failed with the following messages

```
Hello! We are building CUDA driver for your Linux kernel ...
Cloning into 'linux'...
Note: checking out '92b8a3b2136d3051170ed6b14f21bac7808acc09'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b new_branch_name

Checking out files: 100% (52916/52916), done.
Switched to a new branch 'stable'
M   include/uapi/linux/netfilter/xt_CONNMARK.h
M   include/uapi/linux/netfilter/xt_DSCP.h
M   include/uapi/linux/netfilter/xt_MARK.h
M   include/uapi/linux/netfilter/xt_RATEEST.h
M   include/uapi/linux/netfilter/xt_TCPMSS.h
M   include/uapi/linux/netfilter_ipv4/ipt_ECN.h
M   include/uapi/linux/netfilter_ipv4/ipt_TTL.h
M   include/uapi/linux/netfilter_ipv6/ip6t_HL.h
M   net/netfilter/xt_DSCP.c
M   net/netfilter/xt_HL.c
M   net/netfilter/xt_RATEEST.c
M   net/netfilter/xt_TCPMSS.c
  HOSTCC  scripts/basic/fixdep
  HOSTCC  scripts/kconfig/conf.o
  SHIPPED scripts/kconfig/zconf.tab.c
  SHIPPED scripts/kconfig/zconf.lex.c
  SHIPPED scripts/kconfig/zconf.hash.c
  HOSTCC  scripts/kconfig/zconf.tab.o
  HOSTLD  scripts/kconfig/conf
scripts/kconfig/conf  --silentoldconfig Kconfig
#
# configuration written to .config
#
  SYSTBL  arch/x86/entry/syscalls/../../include/generated/asm/syscalls_32.h
  SYSHDR  arch/x86/entry/syscalls/../../include/generated/asm/unistd_32_ia32.h
  SYSHDR  arch/x86/entry/syscalls/../../include/generated/asm/unistd_64_x32.h
  SYSTBL  arch/x86/entry/syscalls/../../include/generated/asm/syscalls_64.h
  HYPERCALLS arch/x86/entry/syscalls/../../include/generated/asm/xen-hypercalls.h
  SYSHDR  arch/x86/entry/syscalls/../../include/generated/uapi/asm/unistd_32.h
  SYSHDR  arch/x86/entry/syscalls/../../include/generated/uapi/asm/unistd_64.h
  SYSHDR  arch/x86/entry/syscalls/../../include/generated/uapi/asm/unistd_x32.h
  HOSTCC  scripts/basic/bin2c
  HOSTCC  arch/x86/tools/relocs_32.o
  HOSTCC  arch/x86/tools/relocs_64.o
  HOSTCC  arch/x86/tools/relocs_common.o
  HOSTLD  arch/x86/tools/relocs
  CHK     include/config/kernel.release
  UPD     include/config/kernel.release
  WRAP    arch/x86/include/generated/asm/clkdev.h
  WRAP    arch/x86/include/generated/asm/cputime.h
  WRAP    arch/x86/include/generated/asm/dma-contiguous.h
  WRAP    arch/x86/include/generated/asm/early_ioremap.h
  WRAP    arch/x86/include/generated/asm/mcs_spinlock.h
  WRAP    arch/x86/include/generated/asm/mm-arch-hooks.h
  CHK     include/generated/uapi/linux/version.h
  UPD     include/generated/uapi/linux/version.h
  CHK     include/generated/utsrelease.h
  UPD     include/generated/utsrelease.h
  CC      arch/x86/purgatory/purgatory.o
  AS      arch/x86/purgatory/stack.o
  AS      arch/x86/purgatory/setup-x86_64.o
  CC      arch/x86/purgatory/sha256.o
  AS      arch/x86/purgatory/entry64.o
  CC      arch/x86/purgatory/string.o
  LD      arch/x86/purgatory/purgatory.ro
  BIN2C   arch/x86/purgatory/kexec-purgatory.c
  CC      kernel/bounds.s
  CHK     include/generated/bounds.h
  UPD     include/generated/bounds.h
  CHK     include/generated/timeconst.h
/bin/sh: 1: bc: not found
make[1]: *** [include/generated/timeconst.h] Error 127
make: *** [prepare0] Error 2
patching file /opt/share/cuda/NVIDIA-Linux-x86_64-352.39/kernel/nv-procfs.c
Hunk #2 succeeded at 563 (offset -1 lines).

ERROR: Unable to find the module utility `insmod`; please make sure you
       have the package 'module-init-tools' or 'kmod' installed.  If you do
       have 'module-init-tools' or 'kmod' installed, then please check that
       `insmod` is in your PATH.

cp: cannot stat `/opt/share/cuda/NVIDIA-Linux-x86_64-352.39/kernel/uvm/nvidia-uvm.ko': No such file or directory
cp: cannot stat `/opt/share/cuda/NVIDIA-Linux-x86_64-352.39/kernel/nvidia.ko': No such file or directory
```

wangkuiyi commented:
我也碰到这个问题了。为了找到原因，我 vagrant ssh 后 

```
docker run -it --rm -v /home/core/share:/opt/share --privileged ubuntu:14.04 /bin/bash
```

随后单步执行 `build-cuda-driver.sh` 里的命令。这样在

```
/opt/share/cuda/NVIDIA-Linux-x86_64-352.39/nvidia-installer -q -a -n -s --kernel-source-path=/opt/share/linux/
```

之后，我有机会查看build log：

```
less /var/log/nvidia-installer.log
```

看到一些编译错误：

```
 /opt/share/cuda/NVIDIA-Linux-x86_64-352.39/kernel/conftest/functions.h:13:2: error: #error kmem_cache_create() conftest failed!
    #error kmem_cache_create() conftest failed!
     ^
   /opt/share/cuda/NVIDIA-Linux-x86_64-352.39/kernel/conftest/functions.h:14:2: error: #error on_each_cpu() conftest failed!
    #error on_each_cpu() conftest failed!
     ^
   /opt/share/cuda/NVIDIA-Linux-x86_64-352.39/kernel/conftest/functions.h:15:2: error: #error smp_call_function() conftest failed!
    #error smp_call_function() conftest failed!
     ^
   /opt/share/cuda/NVIDIA-Linux-x86_64-352.39/kernel/conftest/functions.h:20:2: error: #error acpi_walk_namespace() conftest failed!
    #error acpi_walk_namespace() conftest failed!
     ^
   /opt/share/cuda/NVIDIA-Linux-x86_64-352.39/kernel/conftest/functions.h:22:2: error: #error pci_dma_mapping_error() conftest failed!
    #error pci_dma_mapping_error() conftest failed!
```

这是之前我没有看到过的。但是和这篇5月10日的帖子描述的一样：https://devtalk.nvidia.com/default/topic/935145/makefile-errors-building-nvidia-module-with-nvidia-352-on-debian-jessie/?offset=1

我怀疑之前我都是在CoreOS stable上试验。最近把Vagrantfile里的CoreOS channel改成alpha了。我试试改回stable。

wangkuiyi commented:
我用stable的时候，连Linux kernel source code的git clone命令都不行了。碰到的问题和 @lipeng-unisound  之前碰到的一样：git clone 执行过程中停在那不动了。但是如果不在Docker container里执行同样的git clone命令，是可以下载源码的。在interactive Docker container里，也可以下载。只是 Docker container 里调用 git clone 命令的时候，会停。

lipeng-unisound commented:
@wangkuiyi 我们描述的可能不是一个问题。我 vagrant ssh 后，

``` bash
core@core-01 ~ $ docker run -it --rm -v /home/core/share:/opt/share gcc:4.9 /bin/bash
```

然后在 docker container 里单步执行 build-cuda-driver.sh 中的命令，执行到 make modules_prepare 时得到下面的错误

``` bash
root@9ae5ee877bc3:/opt/share/linux# make modules_prepare
scripts/kconfig/conf  --silentoldconfig Kconfig
#
# configuration written to .config
#
  CHK     include/config/kernel.release
  CHK     include/generated/uapi/linux/version.h
  CHK     include/generated/utsrelease.h
  UPD     include/generated/utsrelease.h
  CHK     include/generated/bounds.h
  CHK     include/generated/timeconst.h
/bin/sh: 1: bc: not found
make[1]: *** [include/generated/timeconst.h] Error 127
make: *** [prepare0] Error 2
```

所以我怀疑是 gcc:4.9 个 docker image 中没有包含必要的工具？

Commit:
Install missing tools in the gcc:4.9 docker image

The procedure of building CUDA driver needs bc and insmod which are
absent in the gcc:4.9 docker image, so we need to install them in the
shell script.

lipeng-unisound commented:
我在 build-cuda-driver.sh 加上了安装两个必需工具的的命令，现在可以正常生成 nvidia.ko 和 nvidia-uvm.ko 两个文件了。

以下更新已经 push 到 lipeng-unisound/fix 

``` diff
diff --git a/build-cuda-driver.sh b/build-cuda-driver.sh
index 245af8a..7c0a491 100755
--- a/build-cuda-driver.sh
+++ b/build-cuda-driver.sh
@@ -7,6 +7,10 @@ if [[ `gcc --version | head -n 1` != "gcc (GCC) 4.9.3" ]]; then
     exit;
 fi

+# Install the essential tools
+apt-get update
+apt-get install -y bc module-init-tools
+
 # clone Linux kernel source code and prepare for kernel module building.
 cd /opt/share
 git clone -b v`uname -r | sed -e "s/-.*//" | sed -e "s/\.[0]*$//"` --depth 1 git://git.kernel.org/pub/scm/linux/kernel/git/stable/linux-stable.git linux
```

Commit:
Merge pull request #6 from k8sp/lipeng-unisound/fix

lipeng-unisound/fix
